dynamic_programming/cut_rod.cpp

```diff
@@ -1,25 +1,104 @@
-/*Given a rod of length n inches and an array of prices that
-contains prices of all pieces of size smaller than n. Determine
-the maximum value obtainable by cutting up the rod and selling
-the pieces.*/
+/**
+ * @file
+ * @brief Implementation of cutting a rod problem
+ *
+ * @details
+ * Given a rod of length n inches and an array of prices that
+ * contains prices of all pieces of size<=n. Determine
+ * the maximum profit obtainable by cutting up the rod and selling
+ * the pieces.
+ *
+ * ### Algorithm
+ * The idea is to break the given rod into every smaller piece as possible
+ * and then check profit for each piece, by calculating maximum profit for
+ * smaller pieces we will build the solution for larger pieces in bottom-up
+ * manner.
+ *
+ * @author [Anmol](https://github.com/Anmol3299)
+ * @author [Pardeep](https://github.com/Pardeep009)
+ *
+ */
 
+#include <array>
+#include <cassert>
+#include <climits>
 #include <iostream>
-using namespace std;
-int cutrod(int p[], int n) {
-    int r[n + 1];
-    r[0] = 0;
-    for (int j = 0; j < n; j++) {
+/**
+ * @namespace dynamic_programming
+ * @brief Dynamic Programming algorithms
+ */
+namespace dynamic_programming {
+/**
+ * @namespace cut_rod
+ * @brief Implementation of cutting a rod problem
+ */
+namespace cut_rod {
+/**
+ * @brief maxProfitByCuttingRod function cuts the rod in different pieces and
+ * stores the maximum profit for each piece of the rod.
```

Comment:
Please document what the template parameter `T` does.
```suggestion
 * stores the maximum profit for each piece of the rod.
 * @tparam T description
```